subcommand/connect-init/command.go

```diff
@@ -140,6 +150,19 @@ func (c *Command) Run(args []string) int {
 		}
 		for _, svc := range serviceList {
 			c.UI.Info(fmt.Sprintf("Registered service has been detected: %s", svc.Service))
+			if c.flagACLAuthMethod != "" {
+				if c.flagServiceName != "" && c.flagServiceAccountName != c.flagServiceName {
+					// Set the error but return nil so we don't retry.
+					errServiceNameMismatch = fmt.Errorf("service account name %s doesn't match annotation service name %s", c.flagServiceAccountName, c.flagServiceName)
+					return nil
+				}
+
+				if c.flagServiceName == "" && c.flagServiceAccountName != svc.Meta[connectinject.MetaKeyKubeServiceName] {
```

Comment:
I think it'd be better to compare it with the consul service name here to make it a bit more obvious even though the two names should be equal. This error is specifically about consul service name != service account name:
```suggestion
				if c.flagServiceName == "" && c.flagServiceAccountName != svc.Name {
```